chain/src/main/scala/org/bitcoins/chain/blockchain/Blockchain.scala

```diff
@@ -53,103 +54,153 @@ case class Blockchain(headers: Vector[BlockHeaderDb])
 
-    val tipResult: BlockchainUpdate = {
-      val nested: Vector[BlockchainUpdate] = blockchains.map { blockchain =>
-        val prevBlockHeaderIdxOpt =
-          blockchain.headers.zipWithIndex.find {
-            case (headerDb, _) =>
-              headerDb.hashBE == header.previousBlockHashBE
-          }
-        prevBlockHeaderIdxOpt match {
-          case None =>
-            logger.warn(
-              s"No common ancestor found in the chain to connect to ${header.hashBE}")
-            val err = TipUpdateResult.BadPreviousBlockHash(header)
-            val failed = BlockchainUpdate.Failed(blockchain = blockchain,
-                                                 failedHeader = header,
-                                                 tipUpdateFailure = err)
-            failed
-
-          case Some((prevBlockHeader, prevHeaderIdx)) =>
-            //found a header to connect to!
-            logger.debug(
-              s"Attempting to add new tip=${header.hashBE.hex} with prevhash=${header.previousBlockHashBE.hex} to chain")
-            val chain = blockchain.fromHeader(prevBlockHeader)
-            val tipResult =
-              TipValidation.checkNewTip(newPotentialTip = header, chain.get)
-
-            tipResult match {
-              case TipUpdateResult.Success(headerDb) =>
-                logger.debug(
-                  s"Successfully verified=${headerDb.hashBE.hex}, connecting to chain")
-                val oldChain =
-                  blockchain.takeRight(blockchain.length - prevHeaderIdx)
-                val newChain =
-                  Blockchain.fromHeaders(headerDb +: oldChain)
-                BlockchainUpdate.Successful(newChain, headerDb)
-              case fail: TipUpdateResult.Failure =>
-                logger.warn(
-                  s"Could not verify header=${header.hashBE.hex}, reason=$fail")
-                BlockchainUpdate.Failed(blockchain, header, fail)
-            }
+    val tipResult: ConnectTipResult = {
+      val prevBlockHeaderIdxOpt =
+        blockchain.headers.zipWithIndex.find {
+          case (headerDb, _) =>
+            headerDb.hashBE == header.previousBlockHashBE
         }
+      prevBlockHeaderIdxOpt match {
+        case None =>
+          logger.warn(
+            s"No common ancestor found in the chain to connect to ${header.hashBE}")
+          val err = TipUpdateResult.BadPreviousBlockHash(header)
+          val failed = ConnectTipResult.BadTip(err)
+          failed
+
+        case Some((prevBlockHeader, prevHeaderIdx)) =>
+          //found a header to connect to!
+          logger.debug(
+            s"Attempting to add new tip=${header.hashBE.hex} with prevhash=${header.previousBlockHashBE.hex} to chain")
+          val chain = blockchain.fromHeader(prevBlockHeader)
+          val tipResult =
+            TipValidation.checkNewTip(newPotentialTip = header, chain.get)
```

Comment:
Possible to avoid the `.get` here? A `fromValidHeader` method would convey meaning better